plugins/adr/src/components/EntityAdrContent/EntityAdrContent.tsx

```diff
@@ -0,0 +1,156 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useCallback, useEffect, useState } from 'react';
+import { useSearchParams } from 'react-router-dom';
+import usePrevious from 'react-use/lib/usePrevious';
+import {
+  Content,
+  ContentHeader,
+  MissingAnnotationEmptyState,
+  Progress,
+  SupportButton,
+  WarningPanel,
+} from '@backstage/core-components';
+import {
+  AdrFilePathFilterFn,
+  ANNOTATION_ADR_LOCATION_DIR,
+  getAdrLocationUrl,
+  isAdrAvailable,
+  madrFilePathFilter,
+} from '@backstage/plugin-adr-common';
+import { useEntity } from '@backstage/plugin-catalog-react';
+import {
+  Grid,
+  List,
+  ListItem,
+  ListItemText,
+  makeStyles,
+  Theme,
+  Typography,
+} from '@material-ui/core';
+
+import { useOctokitRequest } from '../../hooks';
+import { AdrContentDecorator, AdrReader } from '../AdrReader';
+
+const useStyles = makeStyles((theme: Theme) => ({
+  adrMenu: {
+    backgroundColor: theme.palette.background.paper,
+  },
+}));
+
+/**
+ * Component for browsing ADRs on an entity page.
+ * @public
+ */
+export const EntityAdrContent = ({
+  contentDecorators,
+  filePathFilterFn,
+}: {
+  contentDecorators?: AdrContentDecorator[];
+  filePathFilterFn?: AdrFilePathFilterFn;
+}) => {
+  const classes = useStyles();
+  const { entity } = useEntity();
+  const [adrList, setAdrList] = useState<string[]>([]);
+  const [searchParams, setSearchParams] = useSearchParams();
+  const entityHasAdrs = isAdrAvailable(entity);
+
+  const { value, loading, error } = useOctokitRequest(
+    getAdrLocationUrl(entity),
+  );
+
+  const [selectedAdr, setSelectedAdr] = useState(
+    (searchParams.get('record') ?? '') as string,
+  );
+
+  const selectAdr = useCallback(
+    newAdr => {
+      setSelectedAdr(newAdr);
+      searchParams.set('record', newAdr);
+      setSearchParams(searchParams, { replace: true });
+    },
+    [searchParams, setSearchParams, setSelectedAdr],
+  );
+
+  const prevValue = usePrevious(value);
+  useEffect(() => {
+    if (!value?.data || value === prevValue) {
+      return;
+    }
+
+    const adrs = value.data
+      .filter(
+        (item: { type: string; name: string }) =>
+          item.type === 'file' &&
+          (filePathFilterFn
+            ? filePathFilterFn(item.name)
+            : madrFilePathFilter(item.name)),
+      )
+      .map(({ name }: { name: string }) => name);
+
+    setAdrList(adrs);
+    selectAdr(adrs.includes(selectedAdr) ? selectedAdr : adrs[0]);
```

Comment:
Quite a lot of state being set here, completely optional, but could be worth checking if this is better expressed as for example a reducer. Or if it's possible to extract some of this out from the `useEffect` and compute it more eagerly

Reply:
Can you elaborate? Still picking up React as a go 😅

Reply:
This is calling `setAdrList`, `setSelectedAdr`, and `setSearchParams`. Calling a lot of state setters often causes extra rerenders and also confusion 😅 

Tbh looking at this again now I think a good refactor here could be to remove the separate `selectedAdr` state and drive that using the search params instead. You'd look up the active ADR from the list of ADRs on every render, since that's cheap, and if you fail to find one with the expected name then you could use an effect to switch the search param to the first one in the list. Just being a bit careful about not causing any loops with that :grin:

With that you should also be able to remove the `useCallback` for `selectAdr`, since that exists just for the effect right now.

Reply:
Thanks for the explanation! I made some changes based on your suggestion. Let me know if this is what you had in mind!